File: backend/app/panel/action_rules.py
from __future__ import annotations

import re
from typing import Any

ENTITY_ID = re.compile('^[a-z0-9_]+\\.[a-z0-9_]+$')
VIRTUAL_ENTITY_ID = re.compile('^virtual\\.[a-z0-9_]+\\.[a-z0-9_]+$')
ACTION_TYPES = frozenset({
    'toggle',
    'navigate',
    'more-info',
})
POPUP_SOURCES = frozenset({
    'custom',
    'entity',
    'current',
})
TOGGLE_ENTITY_DOMAINS = frozenset({
    'fan',
    'cover',
    'light',
    'button',
    'remote',
    'script',
    'switch',
    'climate',
    'automation',
    'media_player',
    'water_heater',
    'input_boolean',
})


def valid_ha_entity_id(value: str) -> bool:
    return bool(ENTITY_ID.fullmatch(value))


def valid_entity_id(value: str) -> bool:
    """Accept Home Assistant IDs and the renderer's scoped virtual IDs."""
    return valid_ha_entity_id(value) or bool(VIRTUAL_ENTITY_ID.fullmatch(value))


def action_popup_source(action: Any) -> str:
    data = getattr(action, 'data', None)
    if not isinstance(data, dict) and isinstance(action, dict):
        data = action.get('data')
    source = str((data or {}).get('popupSource') or 'current')
    return source if source in POPUP_SOURCES else 'current'


def action_needs_current_entity(action: Any) -> bool:
    action_type = getattr(action, 'type', None)
    if action_type is None and isinstance(action, dict):
        action_type = action.get('type')
    return action_type == 'toggle' or (
        action_type == 'more-info' and action_popup_source(action) == 'current'
    )


def entity_id_supports_toggle(entity_id: str) -> bool:
    normalized = str(entity_id or '')
    return bool(VIRTUAL_ENTITY_ID.fullmatch(normalized)) or (
        normalized.partition('.')[0] in TOGGLE_ENTITY_DOMAINS
    )

File: frontend/static/js/editor/action-rules.js
import { isVirtualEntityId } from "./virtual-entities.js?v=20260822-icon-visibility-v1";
export const ACTION_TYPES = Object.freeze(["toggle", "more-info", "navigate"]);
export const POPUP_SOURCES = Object.freeze(["current", "entity", "custom"]);
export const TOGGLE_ENTITY_DOMAINS = new Set(["automation", "button", "climate", "cover", "fan", "input_boolean", "light", "media_player", "remote", "script", "switch", "water_heater"]);
export function actionPopupSource(action) {
  const source = String(action?.data?.popupSource || "current");
  if (POPUP_SOURCES.includes(source)) {
    return source;
  } else {
    return "current";
  }
}
export function actionPopupData(action) {
  return {
    source: actionPopupSource(action),
    entityId: String(action?.data?.entityId || ""),
    popupId: String(action?.data?.popupId || "")
  };
}
export function actionNeedsCurrentEntity(action) {
  return action?.type === "toggle" || action?.type === "more-info" && actionPopupSource(action) === "current";
}
export function entityIdSupportsToggle(entityId) {
  const id = String(entityId || "");
  return isVirtualEntityId(id) || TOGGLE_ENTITY_DOMAINS.has(id.split(".", 1)[0]);
}
export function componentActionIsSupported(component, action, {
  pagePaths = null,
  popupIds = null
} = {}) {
  if (!ACTION_TYPES.includes(action?.type) || component?.type === "presence-sensor") {
    return false;
  }
  const entityId = String(component?.bindings?.entity?.entityId || "");
  if (action.type === "toggle") {
    return entityIdSupportsToggle(entityId);
  }
  if (action.type === "navigate") {
    const target = String(action.target || "");
    return !!target && (!pagePaths || pagePaths.has(target));
  }
  const popupSource = actionPopupSource(action);
  if (popupSource === "current") {
    return !!entityId;
  }
  if (popupSource === "entity") {
    return !!action.data?.entityId;
  }
  const popupId = String(action.data?.popupId || "");
  return !!popupId && (!popupIds || popupIds.has(popupId));
}
